ext/sodium/libsodium.c

```diff
@@ -0,0 +1,2682 @@
+							  &msg, &msg_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (key_len != crypto_shorthash_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_shorthash(): key size should be "
+				   "CRYPTO_SHORTHASH_KEYBYTES bytes",
+				   0);
+		return;
+	}
+	hash = zend_string_alloc(crypto_shorthash_BYTES, 0);
+	if (crypto_shorthash((unsigned char *) ZSTR_VAL(hash), msg,
+						 (unsigned long long) msg_len, key) != 0) {
+		zend_string_free(hash);
+		zend_throw_exception(sodium_exception_ce, "crypto_shorthash(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(hash)[crypto_shorthash_BYTES] = 0;
+
+	RETURN_STR(hash);
+}
+
+PHP_FUNCTION(sodium_crypto_secretbox)
+{
+	zend_string	  *ciphertext;
+	unsigned char *key;
+	unsigned char *msg;
+	unsigned char *nonce;
+	size_t		   key_len;
+	size_t		   msg_len;
+	size_t		   nonce_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss",
+							  &msg, &msg_len,
+							  &nonce, &nonce_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (nonce_len != crypto_secretbox_NONCEBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox(): nonce size should be "
+				   "CRYPTO_SECRETBOX_NONCEBYTES bytes",
+				   0);
+		return;
+	}
+	if (key_len != crypto_secretbox_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox(): key size should be "
+				   "CRYPTO_SECRETBOX_KEYBYTES bytes",
+				   0);
+		return;
+	}
+	if (SIZE_MAX - msg_len <= crypto_secretbox_MACBYTES) {
+		zend_throw_exception(sodium_exception_ce, "arithmetic overflow", 0);
+		return;
+	}
+	ciphertext = zend_string_alloc((size_t) msg_len + crypto_secretbox_MACBYTES, 0);
+	if (crypto_secretbox_easy((unsigned char *) ZSTR_VAL(ciphertext),
+							  msg, (unsigned long long) msg_len,
+							  nonce, key) != 0) {
+		zend_string_free(ciphertext);
+		zend_throw_exception(sodium_exception_ce, "crypto_secretbox(): internal error", 0);
+		return;
+	}
+	ZSTR_VAL(ciphertext)[msg_len + crypto_secretbox_MACBYTES] = 0;
+
+	RETURN_STR(ciphertext);
+}
+
+PHP_FUNCTION(sodium_crypto_secretbox_open)
+{
+	zend_string	  *msg;
+	unsigned char *key;
+	unsigned char *ciphertext;
+	unsigned char *nonce;
+	size_t		   key_len;
+	size_t		   ciphertext_len;
+	size_t		   nonce_len;
+
+	if (zend_parse_parameters(ZEND_NUM_ARGS(), "sss",
+							  &ciphertext, &ciphertext_len,
+							  &nonce, &nonce_len,
+							  &key, &key_len) == FAILURE) {
+		return;
+	}
+	if (nonce_len != crypto_secretbox_NONCEBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox_open(): nonce size should be "
+				   "CRYPTO_SECRETBOX_NONCEBYTES bytes",
+				   0);
+		return;
+	}
+	if (key_len != crypto_secretbox_KEYBYTES) {
+		zend_throw_exception(sodium_exception_ce,
+				   "crypto_secretbox_open(): key size should be "
+				   "CRYPTO_SECRETBOX_KEYBYTES bytes",
+				   0);
+		return;
+	}
+	if (ciphertext_len < crypto_secretbox_MACBYTES) {
+		RETURN_FALSE;
```

Comment:
The double error handling looks fine to me. They cover very different cases. The former is not an error. And most people will not even try to catch the exception, which is fine as well, as it will only be thrown on misuse.